Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java

```diff
@@ -1342,12 +1340,12 @@ private void updateInformationFromServer(DocumentEvent e) {
         final int eventOffset = e.getOffset();
         final int deltaLength = e.getNewLength() - e.getOldLength();
         for (PluginNavigationRegion region : regions) {
-          if (region.offset <= eventOffset && eventOffset <= region.offset + region.length) {
-            region.length += deltaLength;
-          }
-          else if (region.offset >= eventOffset) {
+          if (eventOffset <= region.offset) {
             region.offset += deltaLength;
           }
+          else if (region.offset < eventOffset && eventOffset <= region.offset + region.length) {
+            region.length += deltaLength;
```

Comment:
@scheglov With typing inside the region most likely resolution target will change. I think removing the region would be better, would you agree?

Also some more cases are not handled. When selection is deleted in the editor DocumentEvent range and region may intersect, for example:
`eventOffset < region.offset < eventOffset + eventLength < region.offset + region.length`
`region.offset < eventOffset < region.offset + region.length < eventOffset + eventLength`
Will you agree that in case of any intersection we should delete the region from `myNavigationData`.

I noticed this when was working on highlighting. I'm going to add the same logic for `myHighlightData`.